### PR TITLE
[WIP] ML Workflow

### LINKED ISSUE #17: Dummy ML workflow
In addition to #16, a dummy ML workflow could be useful, showing how to merge files per dataset and use merged files of multiple datasets to a "bottleneck" task such as a ML training. Also, we could implement some deterministic, n-fold event splitting for cross-validation later on. The inference part would be identical to #16.

### PR DESCRIPTION
**Pull request by riga**
This PR adds all objects needed to perform ML trainings and evaluations. This required a few generalizations of existing, non-ML-related code:

- Existing producers/calibrators/selectors should be called while bound to the instance, e.g. `def my_producer(events, **kwargs)` becomes `def my_producer(self, events, **kwargs)`. This is far more consistent and provides access to the instance itself.
- `deterministic_seeds` is now a producer as it should be (was a calibrator before). Also, it's fixed for use with real data.
- Some minor code cleanup here and there. Pardon me :)

The new ML-related changes are:

- Added a `venv_ml_tf[_dev]` sandbox. Python 3.6 forces us to use TF 2.6.
- Added a new `MLModel` interface with is a super basic layer that defines model training and evaluation, plus some connections to the task and the usual `uses/produces` columns.
- A dummy model that implements the interface for testing.
- Two new mixins: `MLModelMixin` and `MLModelsMixin`. This fits nicely into the existing mixin-logic and is mostly trivial.
- Four new tasks that perform the workflow and that all inherit from `MLModelMixin` (singular):
  - `PrepareMLEvents`: Divides events into folds for k-fold cross validation and writes input features used by the model.
  - `MergeMLEvents`: Merges the output of `PrepareMLEvents` per fold for faster training loops.
  - `MLTraining`: The actual training task producing models.
  - `MLEvaluation`: Performs the evaluation of one model for events seen and unseen during training (k-fold x-validation).
- `CreateHistograms` and all tasks downstream inherit from `MLModelsMixin` (plural) now. If requested, they triggers the training / evaluation of multiple ML models upstream to write their results into histograms.

Closes #17.

The WIP label is due to the dependence on #54 which needs to be merged first before the review makes sense.

### CODE REVIEW

File: ap/columnar_util.py [hunk 1]
```diff
@@ -144,15 +145,32 @@ def check(cls, route: Union["Route", Sequence[str], str]):
...
+        fields. In case no object exists at that index, its values are filled with *null_value*.
+        """
+        # get fields
+        fields = cls.check(route).fields
+
+        # when the last field is an integer, try to get the matching object and maybe pad
+        if len(fields) > 1 and re.match(r"^\d+$", str(fields[-1])):
+            # pad with None
+            idx = int(fields[-1])
+            result = ak.pad_none(ak_array[fields[:-1]], idx + 1, axis=-1)[..., idx]
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/columnar_util.py:166]: E.g. when the route is "Jet.pt.0", idx is 0, and the padding should have the size 1, therefore idx+1.

File: ap/ml/__init__.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        return cls._instances[name]
+
+    @classmethod
+    def get(cls, name: str, copy: bool = False) -> "MLModel":
+        """
+        Returns a previously registered instance named *name* and optionally copies it when *copy*
+        is *True*. A *ValueError* is raised when no instance was found with that name.
+        """
+        if name not in cls._instances:
+            raise ValueError(f"no model named named '{name}' found")
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/ml/__init__.py:93]: Added in the next commit.

File: ap/ml/__init__.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        return "__".join(
+            f"{name}_{self._format_value(value)}"
+            for name, value in self.parameter_pairs(only_significant=True)
+        )
+
+    def parameter_pairs(self, only_significant: bool = False) -> List[Tuple[str, Any]]:
+        """
+        Returns a list of all parameter name-value tuples.
+        """
+        return list(self.parameters.items())
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/ml/__init__.py:156]: I updated the doc string 👍

Significant parameters would be those that can change the network results (e.g. parameters that control the NN architecture), as opposed to insignificant ones (e.g. some log level during training).

File: ap/ml/test.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ak = maybe_import("awkward")
+tf = maybe_import("tensorflow")
+
+
+class TestModel(MLModel):
+
+    def set_config(self, *args, **kwargs):
+        super().set_config(*args, **kwargs)
+
+        # dynamically add variables for the quantities produced by thia model
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/ml/test.py:25]: Changed in the next commit.

File: ap/ml/test.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        return {"ht", "n_jet", "n_muon", "n_electron", "normalization_weight"}
+
+    def produces(self) -> Set[Union[Route, str]]:
+        return {f"{self.name}.n_muon", f"{self.name}.n_electron"}
+
+    def output(self, task: law.Task) -> FunctionArgs:
+        return FunctionArgs(f"mlmodel_f{task.fold}of{self.folds}", dir=True)
+
+    def open_model(self, output: law.LocalDirectoryTarget) -> tf.keras.models.Model:
+        return tf.keras.models.load_model(output.path)
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/ml/test.py:56]: True. I renamed it to just "target" in the next commit.

File: ap/ml/test.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        self,
+        task: law.Task,
+        input: Any,
+        output: law.LocalDirectoryTarget,
+    ) -> None:
+        # define a dummy NN
+        x = tf.keras.Input(shape=(2,))
+        a1 = tf.keras.layers.Dense(10, activation="elu")(x)
+        y = tf.keras.layers.Dense(2, activation="softmax")(a1)
+        model = tf.keras.Model(inputs=x, outputs=y)
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/ml/test.py:68]: I would consider the `test.py` file already external :)
But I guess we will have to see how this behaves and scales in the future. Perhaps it's okay for this test model in the PR.

File: ap/tasks/framework/mixins.py [hunk 7]
```diff
@@ -387,12 +381,111 @@ def producer_funcs(self):
...
+
+        # get the ML model instance
+        self.ml_model_inst = self.get_ml_model_inst(self.ml_model, self.config_inst)
+
+    def events_used_in_training(self, dataset_inst: od.Dataset, shift_inst: od.Shift) -> bool:
+        # evaluate whether the events for the combination of dataset_inst and shift_inst
+        # shall be used in the training
+        return (
+            dataset_inst in self.ml_model_inst.used_datasets and
+            not shift_inst.x("disjoint", False)
```
[the remark this answers is not in the document]
Comment (riga) [reply ap/tasks/framework/mixins.py:441]: Yeah, should indeed be more descriptive. With `disjoint` I wanted to refer to the fact that the set of produced events related to these shifts are fully disjoint from the nominal ones.

It's called `disjoint_from_nominal` now :)

**Comment by riga** [comment]
Thanks for the review @pkausw !
Feel free to resolve and merge if you're fine with the responses.

File: ap/ml/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pkausw) [reply ap/ml/__init__.py]: I see, thanks :+1:

File: ap/ml/__init__.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    .. py:attribute:: folds
+       type: int
+
+       The number of folds for the k-fold cross-validation.
+
+    .. py:attribute:: parameters
+       type: OrderedDict
+
+       A dictionary mapping parameter names to arbitrary values.
```
[the remark this answers is not in the document]
Comment (pkausw) [reply ap/ml/__init__.py]: Ahhhh, ok :+1: Right, having examples for the parameters might help, especially in the doc string :+1:

**Review by pkausw — APPROVED** [review]
